j5/backends/hardware/sr/v4/serial/protocol.py

```diff
@@ -0,0 +1,170 @@
+
+        :param version: The version string.
+        :returns: A parsed version string.
+        :raises ValueError: The version string was not valid.
+        """
+        match = re.match(r"^(\d+)\.(\d+)(?:\.(\d+))?$", version)
+        if not match:
+            raise ValueError(f"Version does not match format: {version}")
+
+        version_parts = (int(x) for x in match.groups(0))
+        return BoardVersion(*version_parts)
+
+    def _check_firmware_version_supported(
+        self,
+        version_string: str,
+        minimum_fw_version: int = 4,
+    ) -> None:
+        """
+        Raises an exception if the firmware version is not supported.
+
+        :param version_string: the firmware version string to check.
+        :param minimum_fw_version: the lowest supported major firmware version
+        :raises CommunicationError: the board is running unsupported firmware
+        """
+        try:
+            version = self._parse_version_string(version_string)
+        except ValueError as e:
+            raise CommunicationError(f"Unable to parse version number: {version}") from e
+
+        if version.hw_ver != 4:
+            raise CommunicationError(
+                f"Expected hardware version number to be 4, got {version.hw_ver}",
+            )
+
+        if version.fw_major < minimum_fw_version:
+            raise CommunicationError(
+                f"Expected major version number of at least {minimum_fw_version},"
+                f" got {version.fw_major}",
+            )
+
+    def request(self, command: str) -> Optional[str]:
+        """
+        Sends a request to the board.
+
+        :param command: The command to be sent as part of the request.
+        :returns: Response, if any.
+        :raises CommunicationError: Failed to send request.
+        """
+        request_data = command.encode("ascii") + b'\n'
+
+        with self._lock:
+            try:
+                bytes_written = self._serial.write(request_data)
+                if len(request_data) != bytes_written:
+                    raise CommunicationError(
+                        "Mismatch in command bytes written to serial interface.",
+                    )
+            except SerialTimeoutException as e:
+                raise CommunicationError(f"Serial Timeout Error: {e}") from e
+            except SerialException as e:
+                raise CommunicationError(f"Serial Error: {e}") from e
+
+            response = self.read_serial_line()
+
+        if response[:4] == "NACK":
+            _, error_string = response.split(":", 1)
+            raise CommunicationError(f"Power Board returned an error: {error_string}")
+
+        expects_response = command[-1:] == "?"
+        if expects_response:
+            return response
+        elif response == "ACK":
+            return None
+        else:
+            raise CommunicationError("Expected ACK from Power Board,"
+                                     f"but got: {response}")
```

Comment:
This duplicates the functionality of the caller using query.

Reply:
d5572689db1922ca334f3b67c76d29f555a8a819